### DISCUSSION TITLE
Cannot build Adonis.js API in Bitbucket CI/CD Pipeline

### DISCUSSION THREAD

**Discussion by FrenchMajesty** [adonisjs/core#2143] [Help] [answered]
[bitbucket-pipeline.txt](https://github.com/adonisjs/core/files/5826893/bitbucket-pipeline.txt)

Like this title suggests, when running `npm run build` the pipeline fails with the following error:

```
> node ace build --production
[ error ]  "build" command not found
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! practice-os-backend@1.0.0 build: `node ace build --production`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the practice-os-backend@1.0.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

I tried to add a step to output the versions and here's what I get back:

```
node -v && npm -v && node ace -v && tsc -v
v12.19.0
6.14.8
╭───────────────────────────────────────────────────╮
│    node ace --version                             │
│───────────────────────────────────────────────────│
│                                                   │
│    App version: 1.0.0                             │
│    Framework version: 5.0.4-preview-rc-2.1        │
│    Assembler version: Not Installed               │
│                                                   │
╰───────────────────────────────────────────────────╯
tsc command not found
```

The build steps runs fine on local obviously... (never that simple) However, when I run the same version check in local I notice that "Assembler version" is `3.0.6` as opposed to `Not Installed`. I don't know why that is the case or how to fix this error? Overall I just want to be able to build my application in the pipeline to deploy it but this has got me stuck.

-----

The second issue I've noticed is that I get the error:
```
tsc command not found
```

Even though I have a step to install `typescript` globally. Here's the log output for that particular section:
```
 + npm install -g typescript
/nodejs/bin/tsserver -> /nodejs/lib/node_modules/typescript/bin/tsserver
/nodejs/bin/tsc -> /nodejs/lib/node_modules/typescript/bin/tsc
+ typescript@4.1.3
added 1 package from 1 contributor in 3.904s
```
I have attached my `bitbucket-pipeline.yml` file. Any help would be greatly appreciated. Thank you the community :)

**Comment by FrenchMajesty** [comment]
Update: following this[ blog post](https://nozzlegear.com/blog/everything-that-went-wrong-when-deploying-my-node-hapi-app-to-azure) I figured it made sense that my `devDependencies` (which is what typescript is) wouldn't install so I tried to force install all of them with the following steps:
```sh
- npm install --unsafe-perm
- npm install --only=prod
- npm install --only=dev
```

However, after adding this, I still get ` tsc command not found` in my pipeline logs. I have attached the Bitbucket logs for further examination

[pipelineLog-9.txt](https://github.com/adonisjs/core/files/5826887/pipelineLog-9.txt)

**Comment by thetutlage** [comment]
I am not sure why the following install commands are required

```
            - npm install -g typescript
            - npm install --unsafe-perm
            - npm install --only=prod
            - npm install --only=dev

```

It should be simply `npm install` inside the project root

**Reply by FrenchMajesty** [reply]
@thetutlage They're not required. I added them as I was debugging and I didn't remove them as to not distract from the "have you tried this other variation of `npm install` ?" (to show I have indeed tried)

**Reply by thetutlage** [reply]
Can you try with just `npm install` and `node ace build --production` and then share the output?

**Reply by FrenchMajesty** [reply]
Here is my [package.json](https://github.com/adonisjs/core/files/5837054/package.json.txt) file for further inspection

`bitbucket-pipeline.yaml` file:

```yaml
      - step:
            name: Install dependencies
            caches:
              - node
            script:
              - npm install
      - parallel:
        - step:
            name: Build App
            caches:
              - node
            script:
              - npm run build
```

Here is the logs:

```
> node ace build --production
[ error ]  "build" command not found
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! practice-os-backend@1.0.0 build: `node ace build --production`
npm ERR! Exit status 1
npm ERR! 
```

**Reply by thetutlage** [reply]
I see you are running the most recent version of `@adonisjs/core`, but an older version of the `@adonisjs/assembler`. Please upgrade it to the recent alpha, ie `3.1.1`

```sh
npm i -D @adonisjs/assembler@alpha
```

**Reply by FrenchMajesty** [reply]
I updated the version but still same error.

**Reply by thetutlage** [reply]
Are you able to build the project on your local?

**Reply by FrenchMajesty** [reply]
Yes, I mentioned in the original post that there is no build problem locally. Only in pipeline environment.

**Reply by thetutlage** [reply]
Yeah, then it's really hard to know that the real issue is. It has something to do with the way dependencies are installed. Is this how your final pipeline file looks like https://github.com/adonisjs/core/files/5826893/bitbucket-pipeline.txt ?. I can give it try myself and see if I spot anything

**Reply by FrenchMajesty** [reply]
You can remove the steps about deploying to production but yes. The bare minimum testable version is:

Try to setup a adonis v5 repo with build step in the pipeline

**Comment by FrenchMajesty** [comment, marked as answer]
I was able to solve this problem by no longer using the Bitbucket Pipeline, but by using the Google Cloud Build. Their CD/CI pipeline product. Our stack is on GCloud so I would advise to look into the alternative provided by your cloud provider which I'm sure Azure and AWS has
